### ISSUE TITLE
How to control a 20 years old Air Conditioner

### ISSUE THREAD

**Issue by ldellus** [arachnetech/homebridge-mqttthing#135]
Hello,

I have a 20 yo air conditioner and using NodeMCU/IRMQTTServer I am able to control it via HTTP and MQTT. For testing purpose I'm using MQTT.fx. The way I control it is by sending a really long payload of 1191 characters containing the IR code. I have a few problems and I looking for help:
1. I was able to read the IR codes from the remote control, but not decode it, so I know what each code is doing, but in one code I have the temperature, AC mode, and Fan mode, so I ended up with 221 different possibilities that covers all the permutations from 18 deg to 30 deg.
2. I can't find a plugin for Homebridge that would allow me to program the 221 possibilities and send the payload to the device. I tried using mqttthing with the Fan type but it allows only On/Off/Speed/Direction. I need more topics.

Any help and direction would be very much appreciated.

Laurent

**Comment by lulek2007** [comment]
Hi,

Maybe Thermostat type will be better ?

or if You use home bridge buy https://www.ibroadlink.com/products/ir-control
and install plugin https://www.npmjs.com/package/homebridge-broadlink-rm 

Regards

Marcin Kowalik

**Comment by ldellus** [comment]
Thanks Marcin, I thought about that and might give it a try. What I really need is a combination of both Fan and Thermostat.

I'll have a look at the Broadlink device.

Cheers,

Laurent

**Comment by iondarie** [comment]
It would be nice it in time AC control would be supported buy this plugin.

I am curently testing my AC with thermostat type but it has it's limitations for example i cannot turn on/off the swing.

**Comment by iondarie** [comment]
I am also interested in controlling my AC with this plugin.

Proper support for ac control/operation would be greatly apreciated.

P.S. tell me where i can send the beer or donations :D

**Comment by arachnetech** [comment]
I think the closest would be:

```
/**
 * Service "Heater Cooler"
 */

Service.HeaterCooler = function(displayName, subtype) {
  Service.call(this, displayName, '000000BC-0000-1000-8000-0026BB765291', subtype);

  // Required Characteristics
  this.addCharacteristic(Characteristic.Active);
  this.addCharacteristic(Characteristic.CurrentHeaterCoolerState);
  this.addCharacteristic(Characteristic.TargetHeaterCoolerState);
  this.addCharacteristic(Characteristic.CurrentTemperature);

  // Optional Characteristics
  this.addOptionalCharacteristic(Characteristic.LockPhysicalControls);
  this.addOptionalCharacteristic(Characteristic.Name);
  this.addOptionalCharacteristic(Characteristic.SwingMode);
  this.addOptionalCharacteristic(Characteristic.CoolingThresholdTemperature);
  this.addOptionalCharacteristic(Characteristic.HeatingThresholdTemperature);
  this.addOptionalCharacteristic(Characteristic.TemperatureDisplayUnits);
  this.addOptionalCharacteristic(Characteristic.RotationSpeed);
};
```

Would that work?

**Comment by iondarie** [comment]
i think so, i can test if needed

**Comment by tobekas** [comment]
HeaterCooler would be the best choice. You must restrict the TargetHeaterCoolerState (min/Max/valid values) to only COOL (=2), if Heating is not possible. You do not have to change this characteristic during operation.

@arachnetech: Can you add HeaterCooler to this plugin?

**Comment by arachnetech** [comment]
Yes, I will take a look.

**Comment by ldellus** [comment]
Thanks guys, looking forward for the updated plugin.

**Comment by arachnetech** [comment]
Re-reading your original description is you’re going to need some pretty complex ‘apply’ functions to map updates relating to the individual elements to your 221 combinations. You might be better off using something like Node Red for this part.

Also, the HomeKit HeaterCooler service doesn’t include a light or ‘dry’ mode.

I’ll implement the basic HomeKit service for you to try. I’m wondering whether we also need an extended version of the apply function concept which can combine/split multiple attributes into a single MQTT topic.

**Comment by iondarie** [comment]
I am using IRremoteESP8266 to control my AC, combining multiple attributes would be great in theory but as far as i can see the IRremote library doesnt allow multiple commands on a single mqtt topic. Each function of the AC is mapped to a specific status and cmnd topic.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by ldellus** [comment]
@arachnetech My 20 yo AC doesn't have a light, so not an issue, also Dry mode is not needed. I'll have a look at Node Red, thanks!

**Comment by arachnetech** [comment]
The latest version now has HeaterCooler support.

**Comment by arachnetech** [comment]
Please see latest documentation (with a few fixes to the version published to NPM last night) at https://github.com/arachnetech/homebridge-mqttthing#heater-cooler.

(Thanks for reviewing and spotting, @tobekas!)

**Comment by iondarie** [comment]
thank you for your work.

I have 2 questions maybe you can help.

1. in the main home screen when i hit the AC nothing happens, if i go in the AC and set a mode it still does not do anything. after checking the mqtt topics it seems that IRRemoteESP8266 requires to set power on before you can issue a command.

Is there any way that when issuing a command it can first check that the power is off and if it is first issue a command to power on, then issue the mode command?

2. Is there possible to make the fan speed use a personalized array? i see now it uses numbers but the IRremote library seems to be using values like: auto, min, low, medium, high, max. 

Or if anyone has an ideea for an alternative to the IRremote library i am all ears :D

**Comment by ldellus** [comment]
@arachnetech Thanks a lot, I just got the time now to upgrade the plugin and I'll run some tests over the weekend.

**Comment by ldellus** [comment]
Question: how do I send a payload? I'm trying the following but it doesn't work.
"topics": {
"setTargetHeaterCoolerState": "Carrier/send/'30,38000,8186,1634,470,1820,284,1636,470,163'"
},
I tried different variants with ' and without, nada. I'm obviously missing something.
I can send this payload via MQTT.fx and it works.

**Comment by arachnetech** [comment]
Topics are just strings (unless you use the ‘apply function’ thing - see https://github.com/arachnetech/homebridge-mqttthing/blob/master/README.md#applying-functions-to-mqtt-messages-custom-payload-encodingdecoding for details). The payloads are relatively fixed and configured elsewhere.

**Comment by ldellus** [comment]
Thanks for the quick reply, that makes sense. I kind of guessed that but was not sure. I'll have a look at the apply function.

**Comment by ldellus** [comment]
I can't figure out how to use the apply function, I'm just an EE afterall. Can you please guide me with an example using the above payload?

**Comment by arachnetech** [comment]
What MQTT topic do you want to use for turning it on/off, and what values do you want to use for on and off? If `Carrier/send/` is your topic and `30,38000,8186,1634,470,1820,284,1636,470,163` is your on value, you'd need something like:

```
"topics": {
"setTargetHeaterCoolerState": { "topic": "Carrier/send/", "apply": "return ( message == 'true' ) ? '30,38000,8186,1634,470,1820,284,1636,470,163' : 'off-value-string-here';" },
```

**Comment by ldellus** [comment]
It works!!! Thank you so much!!!
I just want to confirm I understand the logic. The apply function toggles through all the strings. So, in my case, the following topics will each toggle through the different states of my AC:
"topics": {
    "setActive": {
        "topic": "Carrier/send",
        "apply": "return ( message == 'true' ) ? 'Active On string' :  'Active Off string';"
         },
    "setTargetHeaterCoolerState": {
       "topic": "Carrier/send",
       "apply": "return ( message == 'true' ) ? 'State Auto string' : 'State Heat string' :'State Cool string' :'State Off string';"
         },
Now, my question is, how can I keep each string in a file under the .homebridge folder, so my config file stays nice a readable?

**Comment by arachnetech** [comment]
Actually, that’s not quite right...

The `?` and `:` there are a special syntax available in JavaScript (and a few other languages), used like `<expression> ? <true_value> : <false_value>` and evaluating to `<true_value>` if `<expression>` is true, otherwise to `<false_value>`. I’m using this to map between `true` and anything other than `true` to your on value and off value.

To map from a series of input values to a series of output values, I’d suggest using something like this:

```
var sourceVals = [ ‘AUTO’, ‘HEAT’, ‘COOL’ ];
var destVals = [ ‘your_auto_val’, ‘your_heat_val’, ‘your_cool_val’ ];
return destVals[ sourceVals.indexOf( message ) ];
```

... which finds the index of the `message` value in the `sourceVals` array and returns the corresponding item from `destVals`. For easier use in a configuration file you can shorten this to a single line as:

`return [ 'your_auto_val', 'your_heat_val', 'your_cool_val' ][ [ 'AUTO', 'HEAT', 'COOL' ].indexOf( message ) ];`

**Comment by arachnetech** [comment]
If you want to use the same custom values for `getTargetHeaterCoolerState` messages (or you don’t intend to use `getTargetHeaterCoolerState` at all), then you could just configure your values in `targetHeaterCoolerValues` of course - then Homebridge-mqttthing will use your values directly without the need for an `apply()` function!

**Comment by ldellus** [comment]
Thanks, I'll have a look at that.

**Comment by ldellus** [comment]
Hi,

My apologies for not updating the forum. I got it working perfectly well via Homebridge and the Homebridge Mqtt Acpartner plugin after crankyoldgit developed the code I needed for IRremoteESP8266.

Thanks,

Laurent
